src/lib/raft/raft.c

```diff
@@ -695,14 +721,51 @@ raft_worker_handle_io(struct raft *raft)
 				raft_sm_wait_election_end(raft);
 			} else {
 				/* No leaders, no votes. */
-				raft_sm_schedule_new_vote(raft, raft->self);
+				raft_sm_schedule_new_vote(raft, raft->self,
+							  raft->vclock);
 			}
 		}
 	} else {
 		memset(&req, 0, sizeof(req));
 		assert(raft->volatile_term >= raft->term);
-		req.term = raft->volatile_term;
+		/*
+		 * Vote and term bumps are persisted separately for foreign
+		 * nodes. This serves as a flush of all transactions going to
+		 * WAL right now so as the local node could correctly compare
+		 * its own vclock vs candidate's one. Otherwise the local vclock
+		 * can be <= candidate's now, but that can change after the WAL
+		 * queue is flushed.
+		 */
+		if (raft->volatile_vote == raft->vote ||
+		    raft->volatile_vote == 0)
+			goto do_dump_with_vote;
+		/*
+		 * When vote for self, it is always acceptable. Can persist it
+		 * together with the term.
+		 */
+		if (raft->volatile_vote == raft->self)
```

Comment:
Maybe unify voting for self and for someone else?
Otherwise raft behaves differently based on some specifics not related to raft.
I think we could always persist term and vote separately

Reply:
This makes the patch notably more complicated mostly due to much bigger diff and because users start seeing intermediate states on the current node via `on_election` callbacks. Also I still need to check for self here to avoid calling `raft_can_vote_for()` on own old vclock. Otherwise can fail a check for voting for self and the log about looks really weird. Done anyway, see the new version.